[Navigation] Introduce NavBackStackEntry#provideToCompositionLocals

jossiwolf opened the pull request:
…that provides the NavBackStackEntry to the composition locals needed for all Compose-related navigators.

Relnote: Introduced a new NavBackStackEntry#provideToCompositionLocals that provides the NavBackStackEntry to the relevant composition locals.

## Testing

Test: NavBackStackEntryProviderTest

`NavBackStackEntryProviderTest#testSaveableValueInContentIsSaved` is pretty much just copied from [RememberSaveableTest](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/runtime/runtime-saveable/src/androidAndroidTest/kotlin/androidx/compose/runtime/saveable/RememberSaveableTest.kt) and as always, I'll be very happy about suggestions on improving the tests and additional tests! 

## Issues Fixed

Fixes: b/187229439

jossiwolf committed:
Introduce new NavBackStackEntry#provideToCompositionLocals extension that provides the NavBackStackEntry to the composition locals needed for all Compose-related navigators.

Test: NavBackStackEntryProviderTest
Fixes: b/187229439
Change-Id: If6af74ec67fbedef2cf5d623e9fc44ab6bf76e39

ianhanniballake commented on navigation/navigation-compose/src/main/java/androidx/navigation/compose/NavBackStackEntryProvider.kt:
```diff
@@ -0,0 +1,82 @@
+@SuppressLint("ComposableNaming")
+@Composable
+public fun NavBackStackEntry.provideToCompositionLocals(
+    saveableStateHolder: SaveableStateHolder,
```
Would this work for the simple case if it had a default value of `rememberSaveableStateHolder()`?

jossiwolf replied:
Yeah - I chose not to provide a default value because I felt it might discourage hoisting the `SaveableStateHolder` but I guess that's alright and not a big concern!

andkulikov replied:
I am actually also not sure have a default value here is reasonable. the whole SaveableStateHolder logic is pretty much useless if it is not hoisted somewhere before iterating through all the entries

ianhanniballake replied:
I was thinking about the case in #173 where only the latest destination is shown (multiple bottom sheets at the same time doesn't make much sense) where having this default would make this more self contained.

I'd be fine leaving out the default and having developers have to think about the hoisting of their `SaveableStateHolder` if you think that better.

jossiwolf replied:
We could leave out the default param and specify in the KDocs that the `saveableStateHolder` should be hoisted? Without that remark folks probably wouldn't actually hoist this.

andkulikov replied:
the same SaveableStateHolder should be used for all the backstack entries. with code like this
backStack.forEach {
    it.CompositionLocalProvider(rememberSaveableStateHolder()) {...}
}
we have different holder for each backstack, which is not how it is expected to be used.
so I would vote to remove the default param value just to not add extra confusion (we can explain it with more details in the doc as well)

jossiwolf replied:
@andkulikov Decided to go with `LocalOwnersProvider` as it did make more sense after thinking about it a bit more :) If anybody has any other ideas we can still rename

Also made the param non-optional and improved docs in a8100a0, let me know what you think :)

ianhanniballake commented on navigation/navigation-compose/src/androidTest/java/androidx/navigation/compose/NavBackStackEntryProviderTest.kt:
```diff
@@ -0,0 +1,159 @@
+    // By default, NavBackStackEntrys are in the INITIALIZED state and then get moved to the next
+    // appropriate state by the NavController. In case we aren't testing with a NavController,
+    // this sets the entry's lifecycle state to the passed state so that the entry is active.
+    private fun NavigatorState.createActiveBackStackEntry(
```
With [aosp/1697987](https://android-review.googlesource.com/c/platform/frameworks/support/+/1697987), I think you can refactor this to:
1) Create the `TestNavigator`
2) Attach a `TestNavigatorState`
3) Create a `NavBackStackEntry`
4) `navigate()` to that `NavBackStackEntry`
5) Return the created `NavBackStackEntry` (which will now be in the right state).

jossiwolf replied:
omg yes, awesome, that's exactly what we need here! Thanks!

jossiwolf replied:
I rewrote the tests to follow that so we'll just merge after your CL has been merged; they're obvs failing rn

jossiwolf replied:
Alright - tests are passing! We're good to go once you're happy :)

jossiwolf committed:
Provide a default value for NavBackStackEntry#provideToCompositionLocals' `saveableStateHolder`

Change-Id: I0525a524e8373d9d3d4f7bb11a3c71046a74f64a

jossiwolf committed:
Use TestNavigatorState to set up NavBackStackEntry instead of the ugly createActiveBackStackEntry hack

Change-Id: I8b7101292e107dded8426a7e38f100451720ac95

ianhanniballake commented on navigation/navigation-compose/src/androidTest/java/androidx/navigation/compose/NavBackStackEntryProviderTest.kt:
```diff
@@ -0,0 +1,151 @@
+    val composeTestRule = createComposeRule()
+
+    @Test
+    fun testViewModelStoreOwnerProvided() {
+        val testNavigator = TestNavigator()
+        val testNavigatorState = TestNavigatorState()
+        testNavigator.onAttach(testNavigatorState)
+        val backStackEntry = testNavigatorState.createBackStackEntry(
+            testNavigator.createDestination(),
+            null
+        )
+        testNavigator.navigate(listOf(backStackEntry), null, null)
```
I'd suggest moving all of these lines into a `private fun createBackStackEntry(): NavBackStackEntry` method to avoid the code duplication in this test suite, similar to the [`createNavController()` helper](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-runtime/src/androidTest/java/androidx/navigation/NavControllerTest.kt;l=1994) from other test suites.

jossiwolf committed:
Extract back stack entry creation into test-private method

Change-Id: I8b8c3c03b190e083928d013d82495bc98a9c5a6f

jossiwolf committed:
Add testNonSaveableValueInContentIsNotSaved to save a non-saveable value isn't saved and restored

Change-Id: Ib31cc443174d18912e55ac64e354f6721c4b1bb1

andkulikov commented on navigation/navigation-compose/src/main/java/androidx/navigation/compose/NavBackStackEntryProvider.kt:
```diff
@@ -0,0 +1,82 @@
+        LocalLifecycleOwner provides this,
+        LocalSavedStateRegistryOwner provides this
+    ) {
+        saveableStateHolder.SaveableStateProvider {
```
I think at some point we had a lint rule to not create such proxy lambdas as they adds unnecessary overhead for the compose runtime. meaning that you can just write 
saveableStateHolder.SaveableStateProvider(content)
instead

ianhanniballake commented:
It looks like the `api` files are out of date now with the default parameter. Can you run `updateApi` again?

jossiwolf committed:
Update API after providing default param

Change-Id: I0738fe193a587d057ac88e0766ccf09556681efd

jossiwolf commented:
Oh yeah, sorry for that!

andkulikov commented on navigation/navigation-compose/src/main/java/androidx/navigation/compose/NavBackStackEntryProvider.kt:
```diff
@@ -0,0 +1,82 @@
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+public fun NavBackStackEntry.provideToCompositionLocals(
```
this function name doesn't follow our api guidelines:
the name of the composable function which returns unit should start with a capital letter and be a noun. even if it is an extension/member function on other class. 
similarly to how we have SaveableStateHolder.SaveableStateProvider()
I don't know what is the best name, maybe Ian would have some ideas. something like:
NavBackStackEntry.LocalOwnersProvider()

jossiwolf replied:
Naming is hard :D `LocalOwnersProvider` isn't that descriptive to me, what do you think of `NavBackStackEntry. CompositionLocalProvider`? (What Ian originally suggested, just pascal case) @ianhanniballake

jossiwolf replied:
(btw, AS keeps telling me that it should indeed start with a lowercase letter😂)

ianhanniballake replied:
Yeah, I personally liked the scoping of making this an extension on `NavBackStackEntry` (vs the more obvious top level function of `CompositionLocalProvider(NavBackStackEntry, SaveableStateHolder, @Composable () -> Unit)`) and yet, the capital lettering feels very odd when it comes to using it with extension functions.

jossiwolf replied:
Yeah, I agree - Pascal case extensions are a little weird. I can live with both options but camel case feels more natural to me

andkulikov replied:
I agree that on extension function it looks more strange, but the current guidelines are not having exceptions for extension/member functions. 
with LocalOwners I tried to describe what composition locals would be provided, as with just CompositionLocalProvider name it doesn't explain that. but maybe what exactly we provide is an implementation detail, so ok to me

ianhanniballake replied:
This method would be something that every custom Navigator would use (that's why this needs to be public API).

andkulikov replied:
not blocking this pull-request but just out of my curiosity: what does it mean to have a custom Navigation? if I write my own navigation implementation how will I construct NavBackStackEntry objects?

ianhanniballake replied:
No, a custom [`Navigator`](https://developer.android.com/reference/kotlin/androidx/navigation/Navigator), not custom Navigation. `NavController` was built in a such a way that it knows nothing about Fragments, Compose, Views, Dialogs, etc.

It is the individual `Navigator` instances that are added to the `NavController` that provide that functionality - `ComposeNavigator` is just a first party implementation of the public API surface, just like the implementation in #173 is a Navigator specifically designed for modal bottom sheets. They can be mixed and matched and NavController is the one who multiplexes them correctly and controls the creation of NavBackStackEntry instances for each Navigator.

jossiwolf committed:
Rename provideToCompositionLocals to CompositionLocalProvider to adhere with Compose API guidelines

Change-Id: I2bd33a90236033933168411e70c6fbdb77e0abfe

jossiwolf committed:
Rename CompositionLocalProvider to LocalOwnersProvider and make its saveableStateHolder parameter non-optional

Also improved the docs to make clear that the SaveableStateHolder should be hoisted.

Change-Id: Ie918a80042ba3192cb04114ed9aa8fed0a0e2beb
